lalrpop/src/lr1/lane_table/construct/merge.rs

```diff
@@ -75,78 +75,100 @@ impl<'m, 'grammar> Merge<'m, 'grammar> {
-                       self.context_sets.context_set(state));
-                self.walk(successor)?;
+                }
+                self.walk(successor, group)?
             } else {
-                // search for an existing clone with which we can merge
-                debug!("Merge::walk: union failed, seek existing clone");
-                let existing_clone = {
-                    let context_sets = &mut self.context_sets;
-                    self.clones.get(&successor)
-                               .into_iter()
-                               .flat_map(|clones| clones) // get() returns an Option<Set>
-                               .cloned()
-                               .filter(|&successor1| context_sets.union(state, successor1))
-                               .next()
-                };
-
-                if let Some(successor1) = existing_clone {
-                    debug!("Merge::walk: found existing clone {:?}", successor1);
-                    self.patch_links(state, successor, successor1);
-                    self.walk(successor1)?;
+                debug!("Merge::walk: state without group, context-set = {:?}",
+                    self.groups.context_set_ref(group));
+                // Successor does not belong to any group, so we just try
+                // to merge it into current group
+                if self.groups.merge_state(group, successor, self.rows.get(&successor)) {
+                    self.walk(successor, group)?
                 } else {
-                    // if we don't find one, we have to make a new clone
-                    debug!("Merge::walk: creating new clone of {:?}", successor);
-                    let successor1 = self.clone(successor);
-                    if self.context_sets.union(state, successor1) {
-                        self.patch_links(state, successor, successor1);
-                        self.walk(successor1)?;
-                    } else {
-                        debug!("Merge::walk: failed to union {:?} with {:?}",
-                               state, successor1);
-                        debug!("Merge::walk: state context = {:?}",
-                               self.context_sets.context_set(state));
-                        debug!("Merge::walk: successor context = {:?}",
-                               self.context_sets.context_set(successor1));
-
-                        return Err((self.original_index(state),
-                                    self.original_index(successor1)));
-                    }
+                    debug!("Merge::walk: failed");
+                    return Err((self.original_index(state), self.original_index(successor)))
                 }
             }
         }
-
         Ok(())
     }
 
+    fn split(&mut self, state: StateIndex, group: Group, successor: StateIndex) -> Result<StateIndex, (StateIndex, StateIndex)> {
+        // search for an existing clone with which we can merge
```

Comment:
I don't know how much it matters, but I imagine it would make sense to first search for an existing clone in the current group, *then* do this logic.